### ISSUE TITLE
Fix Adam Optimizer to Implement Paper Correctly

### ISSUE THREAD

**Issue by isaacgerg** [keras-team/keras#10611]
Adam was recently demonstrated to be implemented incorrectly in several packages including Keras.  Propose to fix the optimizer using method described here:

https://arxiv.org/pdf/1711.05101.pdf

**Comment by brge17** [comment]
1. Keras doesn't call it weight decay, it is explicitly called ``l2`` regularization (see https://keras.io/regularizers/) which is indeed accurate.

2. https://arxiv.org/pdf/1705.08292.pdf

**Comment by isaacgerg** [comment]
I don't think that's right.

https://keras.io/optimizers/#adam

   - *decay*: float >= 0. Learning rate decay over each update.



Regarding 2, Im not sure what this has to do with the conversation.  Did
you mean to say that Adam doesnt generalize as well as SGD?  That's besides
the point.

**Comment by brge17** [comment]
Learning rate decay != weight decay.

They just aren't the same thing.

**Comment by prateekgulati** [comment]
Is the optimizer fixed now?
